routing/probability_bimodal.go

```diff
@@ -0,0 +1,238 @@
+	results NodeResults, toNode route.Vertex, amt lnwire.MilliSatoshi,
+	capacity btcutil.Amount) float64 {
+
+	// We first compute the probability for the desired hop taking into
+	// account previous knowledge.
+	directProbability := p.getLiquidityProbability(
+		now, results, toNode, amt, lnwire.NewMSatFromSatoshis(capacity),
+	)
+
+	// The final probability is computed by taking into account other
+	// channels of the node.
+	return p.calculateProbability(directProbability, now, results, toNode)
+}
+
+// getLocalPairProbability computes the probability to reach toNode given a set
+// of previous learnings.
+func (p *probabilityBimodal) getLocalPairProbability(now time.Time,
+	results NodeResults, toNode route.Vertex) float64 {
+
+	return 1.0
+}
+
+// getLiquidityProbability computes the probability to reach a node based
+// on the liquidity distribution in the LN.
+func (p *probabilityBimodal) getLiquidityProbability(now time.Time,
+	results NodeResults, toNode route.Vertex, amt lnwire.MilliSatoshi,
+	capacity lnwire.MilliSatoshi) float64 {
+
+	// Set success and failure amounts.
+
+	// We know that we can send zero satoshis.
+	successAmount := lnwire.MilliSatoshi(0)
+	// We know that we cannot send the full capacity.
+	failAmount := capacity
+
+	result, ok := results[toNode]
+	if ok {
+		if !result.FailTime.IsZero() {
+			// Apply time decay for the amount we can not send.
+			failAmount = cannotSend(
+				result.FailAmt,
+				capacity,
+				now,
+				result.FailTime,
+				p.PenaltyHalfLife,
+			)
+		}
+		if !result.SuccessTime.IsZero() {
+			// Apply time decay for the amount we can send.
+			successAmount = canSend(
+				result.SuccessAmt,
+				now,
+				result.SuccessTime,
+				p.PenaltyHalfLife,
+			)
+		}
+	}
+
+	probability := p.probabilityFormula(
+		float64(capacity),
+		float64(successAmount),
+		float64(failAmount),
+		float64(amt),
+	)
+
+	return probability
+}
+
+// calculateProbability computes the total hop probability based on the channel
+// probability and historic forwarding data of other channels of the node.
+func (p *probabilityBimodal) calculateProbability(directProbability float64,
+	now time.Time, results NodeResults, toNode route.Vertex) float64 {
+
+	// w is a parameter which determines how strongly the other channels of
+	// a node should be incorportated, the higher the stronger.
+	w := p.AprioriWeight
+
+	// hl determines the timeliness of the previous successes/failures
+	// to be taken into account.
+	hl := p.PenaltyHalfLife
+
+	// The channel probability is weighted fully, all other results are
+	// weighted according to how recent the information is.
+	totalProbabilities := directProbability
+	totalWeights := 1.
+
+	for peer, result := range results {
+		// We don't include the direct hop probability.
+		if peer == toNode {
+			continue
+		}
+
+		// We add probabilities weighted by how recent the info is.
+		var weight float64
+		if result.SuccessAmt > 0 {
+			exponent := -float64(now.Sub(result.SuccessTime)) /
```

Comment:
Why switch to decaying successes? Intuitively I always thought that assuming success until proven otherwise is a strong strategy for probability estimation.

Reply:
Tbh, I always just thought of that as a simplifying assumption we could make to ship an initial version sooner. The intuition behind this new distribution is something like: "most of the time the capacity is in one direction or the other, either due to channels starting out like that, or the inherent asymmetric of the sources and sinks in the network (some services only recv, etc)". The approach of never decaying success sort of assumes that we're the only transactors in the network, and that the context/information that led to our past observations never changes/decays. 

As a middle ground, we could experiment with something like a piece-wise approach, where we drop (after decay) to a higher-than-baseline success probability. This would _maybe_ let us tend towards nodes that are at least able to forward regularly. Ultimately, I'm looking forward to finalizing our implementation independent path finding benchmarking framework so we can have somethign more concrete to rest our heuristics/arguments on.

Reply:
This behavior is still there, the algo will still chose successful routes until they are exhausted. I am not 100% convinced of having a symmetric behavior between successes and failures, although I tend to that. One could allow two different decay times and make it a user choice (say longer decay times for successes, shorter decay times for failures - this way we would forget about failures faster?). For now I chose to keep things minimal.